README.md
# Retro - Reflect, Refine, Repeat

![github_workflow](https://github.com/MaibornWolff/retro/actions/workflows/test.yml/badge.svg) ![GitHub release](https://img.shields.io/github/release/yduman/retro.svg?style=flat-square) ![GitHub](https://img.shields.io/github/license/yduman/retro.svg?style=flat-square) [![code style: prettier](https://img.shields.io/badge/code_style-prettier-ff69b4.svg?style=flat-square)](https://github.com/prettier/prettier)

## What is Retro?

Retro is a tool that is used for retrospectives and planning poker sessions. The goal of Retro is to have an easy to use and secure way of having retro and planning poker sessions.

Retro uses a peer-to-peer approach, where all communication including retro and poker information is shared between users only and never reaches a backend server. The backend server is only used for signaling, which is the initial handshake between the participants in a room. This is possible by following the [WebRTC](https://webrtc.org/) protocol.

## Retrospective

![retro_example](./assets/retro_example_page.png)

## Planning Poker

![planning_poker_example](./assets/poker_example_page.png)

## Requirements

- Latest Node.js LTS version
- NPM version >= 7

## Hosting

TBD

## Development

### Install dependencies

```
$ npm install
```

### Run dev servers

```
$ npm run frontend:dev
$ npm run backend:dev
$ npm run watch:shared
```

| Application | Port |
| ----------- | :--: |
| Frontend    | 3000 |
| Backend     | 3001 |
| Signaling   | 3002 |

## Contributing

For information on how to contribute, see [Contributing](https://github.com/MaibornWolff/retro/blob/master/CONTRIBUTING.md)

## Contributors ✨

Thanks goes to these wonderful people ([emoji key](https://allcontributors.org/docs/en/emoji-key)):

<!-- ALL-CONTRIBUTORS-LIST:START - Do not remove or modify this section -->
<!-- prettier-ignore-start -->
<!-- markdownlint-disable -->
<table>
  <tbody>
    <tr>
      <td align="center" valign="top" width="14.28%"><a href="https://github.com/mrpatpat"><img src="https://avatars2.githubusercontent.com/u/2622069?v=4?s=100" width="100px;" alt="Adrian Endrich"/><br /><sub><b>Adrian Endrich</b></sub></a><br /><a href="https://github.com/MaibornWolff/retro/commits?author=mrpatpat" title="Code">💻</a></td>
      <td align="center" valign="top" width="14.28%"><a href="https://github.com/NearW"><img src="https://avatars.githubusercontent.com/u/12533626?v=4?s=100" width="100px;" alt="Ben Willenbring"/><br /><sub><b>Ben Willenbring</b></sub></a><br /><a href="https://github.com/MaibornWolff/retro/commits?author=NearW" title="Code">💻</a> <a href="#ideas-NearW" title="Ideas, Planning, & Feedback">🤔</a> <a href="#maintenance-NearW" title="Maintenance">🚧</a> <a href="#projectManagement-NearW" title="Project Management">📆</a></td>
      <td align="center" valign="top" width="14.28%"><a href="https://github.com/ClaasBusemann"><img src="https://avatars2.githubusercontent.com/u/65392929?v=4?s=100" width="100px;" alt="Claas Busemann"/><br /><sub><b>Claas Busemann</b></sub></a><br /><a href="https://github.com/MaibornWolff/retro/commits?author=ClaasBusemann" title="Code">💻</a></td>
      <td align="center" valign="top" width="14.28%"><a href="https://github.com/sillydomnom"><img src="https://avatars.githubusercontent.com/u/18489820?v=4?s=100" width="100px;" alt="Dominik Schumann"/><br /><sub><b>Dominik Schumann</b></sub></a><br /><a href="https://github.com/MaibornWolff/retro/commits?author=sillydomnom" title="Code">💻</a></td>
      <td align="center" valign="top" width="14.28%"><a href="https://github.com/Hall-Ma"><img src="https://avatars.githubusercontent.com/u/72517530?v=4?s=100" width="100px;" alt="Maria Hallmann"/><br /><sub><b>Maria Hallmann</b></sub></a><br /><a href="https://github.com/MaibornWolff/retro/commits?author=Hall-Ma" title="Code">💻</a> <a href="https://github.com/MaibornWolff/retro/commits?author=Hall-Ma" title="Tests">⚠️</a></td>
      <td align="center" valign="top" width="14.28%"><a href="https://github.com/2mawi2"><img src="https://avatars2.githubusercontent.com/u/17811051?v=4?s=100" width="100px;" alt="Marius Wichtner"/><br /><sub><b>Marius Wichtner</b></sub></a><br /><a href="https://github.com/MaibornWolff/retro/commits?author=2mawi2" title="Code">💻</a></td>
      <td align="center" valign="top" width="14.28%"><a href="https://github.com/PaulaBre"><img src="https://avatars2.githubusercontent.com/u/65403162?v=4?s=100" width="100px;" alt="PaulaBre"/><br /><sub><b>PaulaBre</b></sub></a><br /><a href="https://github.com/MaibornWolff/retro/commits?author=PaulaBre" title="Code">💻</a></td>
    </tr>
    <tr>
      <td align="center" valign="top" width="14.28%"><a href="https://github.com/tobim-dev"><img src="https://avatars3.githubusercontent.com/u/15176413?v=4?s=100" width="100px;" alt="Tobias"/><br /><sub><b>Tobias</b></sub></a><br /><a href="https://github.com/MaibornWolff/retro/commits?author=tobim-dev" title="Code">💻</a></td>
      <td align="center" valign="top" width="14.28%"><a href="https://yduman.github.io/"><img src="https://avatars0.githubusercontent.com/u/11931774?v=4?s=100" width="100px;" alt="Yadullah Duman"/><br /><sub><b>Yadullah Duman</b></sub></a><br /><a href="https://github.com/MaibornWolff/retro/commits?author=yduman" title="Code">💻</a> <a href="https://github.com/MaibornWolff/retro/pulls?q=is%3Apr+reviewed-by%3Ayduman" title="Reviewed Pull Requests">👀</a> <a href="#ideas-yduman" title="Ideas, Planning, & Feedback">🤔</a></td>
      <td align="center" valign="top" width="14.28%"><a href="https://github.com/dostulataa"><img src="https://avatars.githubusercontent.com/u/7762085?v=4?s=100" width="100px;" alt="dostualtaa"/><br /><sub><b>dostualtaa</b></sub></a><br /><a href="https://github.com/MaibornWolff/retro/commits?author=dostulataa" title="Code">💻</a> <a href="#ideas-dostulataa" title="Ideas, Planning, & Feedback">🤔</a> <a href="#maintenance-dostulataa" title="Maintenance">🚧</a> <a href="#projectManagement-dostulataa" title="Project Management">📆</a></td>
      <td align="center" valign="top" width="14.28%"><a href="https://github.com/ravensinth"><img src="https://avatars0.githubusercontent.com/u/1155772?v=4?s=100" width="100px;" alt="ravensinth"/><br /><sub><b>ravensinth</b></sub></a><br /><a href="https://github.com/MaibornWolff/retro/commits?author=ravensinth" title="Code">💻</a></td>
    </tr>
  </tbody>
</table>

<!-- markdownlint-restore -->
<!-- prettier-ignore-end -->

<!-- ALL-CONTRIBUTORS-LIST:END -->

This project follows the [all-contributors](https://github.com/all-contributors/all-contributors) specification. Contributions of any kind welcome!


CHANGELOG.md
# Changelog

## [Unreleased]

### Added

### Changed

### Fixed

### Removed

## [0.6.0] - 2022-03-08

### Added

- peer to peer communication
- participants button and dialog
- waiting list
- requesting access feature

### Changed

- adapted positioning of ui elements
- vote indicator is now smaller and in the upper right corner

### Fixed

- made all existing features work again
- fixed broken import / export functionality

### Removed

## [0.5.0] - 2021-08-07

### Added

- New dark and light theme
- The planning poker now also displays a chart for the voting results. This is handy if you have a large team.
- You can now mark cards as discussed, thanks to @tobim-dev !
- The moderator can now remove participants manually from the planning poker. This is handy if a participant accidentally closed its tab and therefore lost the session.

### Changed

- Improved UI of vote count dialog
- Improved UI of how many votes are left. No more buggy toasts!
- Blurred cards are now displayed as skeletons
- Participants can now only drag there own cards and cannot drag columns. The moderator can still drag all cards and also all columns.

### Fixed

- Deprecated `KeyboardEvent.keyCode`

### Removed

## [0.4.1] - 2020-10-25

### Added

- **Change Planning Poker Units**
  - Previously, Retro was just supporting planning poker for fibonacci numbers from 0 to 34. Now, the moderator can decide between three unit types and their maximum value. The unit types are fibonacci numbers, natural numbers and T-Shirt sizes. The allowed maximum value for the first two is 100 and T-Shirt sizes go from XS to XXL.
- **Usage of zustand**
  - Added the library [zustand](https://github.com/pmndrs/zustand), in order to explore it for the planning poker page. It handles state management really well and will probably be adopted into the board page as well.

### Changed

- Usage of React 17 and ESLint 7

### Fixed

- Multiple fixes regarding TypeScript deployment

### Removed

- `supertest` library from backend, since there was no usage

## [0.4.0] - 2020-09-19

### Added

- **Planning Poker**

  - Retro now provides a page for planning poker sessions
  - The person starting the session is the moderator, the rest joining are participants
  - Every person who joins the session will have its own card where it can provide an estimation for an user story
  - The moderator can set the current user story for estimation by providing the title and optionally an URL to the user story
  - Initially all cards are red, meaning the user did not vote. When a user votes, the card turns green.

- **Blur columns independently**

  - Moderators can now blur columns independently, by clicking the triple dot button on a column
  - Special thanks to [2mawi2](https://github.com/2mawi2) for realizing this feature! ❤️

- **Share session button**
  - The toolbar got an additional button for sharing a session
  - Previously users had to manually copy the URL for sharing the session. This button will now automatically copy the URL to the clipboard when clicked

### Changed

- **Backend refactoring**
  - Backend got refactored to TypeScript
- **Drop yarn**
  - Retro now uses `npm` as the default package manager
- **rmby**
  - Storage clean up is now handled by the `rmby` library

### Fixed

- Styling for buttons on the toolbar

### Removed

- There was a delete endpoint that was being used for tests. Since we will rethink the whole point of testing Retro, we currently don't need this endpoint anymore

## [0.3.5] - 2020-08-04

### Added

- Instead of blurring/unblurring all cards, we are now blurring/unblurring cards that are now owned by an user. Special thanks to [tobim-dev](https://github.com/tobim-dev), [ravensinth](https://github.com/ravensinth) et al. for realizing this feature ❤️
- The greeting text on the homepage now has a typewriter effect

### Removed

- Empty Gatsby Gitbook Template, since we will handle this project on a different repository

## [0.3.4] - 2020-06-24

### Fixed

- Display of multiline text for card content

## [0.3.3] - 2020-06-14

### Added

- Import/Export of board data
  - If you want to reuse a board for another retrospective, you can now export the data of the board as JSON and import it later for another session
  - Special thanks to [ClaasBusemann](https://github.com/ClaasBusemann) and [PaulaBre](https://github.com/PaulaBre) for their contribution! ❤️
- Docker development support
- Container-based deployment support
- Rate-limiting API
- Configurable CORS settings in PROD mode

### Changed

- Frontend
  - Using TypeScript now!
  - Major UI redesign!
  - Hooks all the way - no more HOCs!
  - Increased maximum card content length
- Backend
  - Storage clean up will now be handled via Node.js instead of writing your own cronjob manually

### Fixed

- The call to `getBoard()` on the backend now handles parsing errors and redirects the user to the error page

### Removed

## [0.3.2] - 2019-10-21

### Added

- Retro Formats
  - On board creation, the user can now select well known retrospective formats

### Changed

- Maximum card content length is now 10000

## [0.3.1] - 2019-09-30

### Changed

- Some dependency updates
- Dialogs don't display an error state by default (#44)

### Fixed

- Fixed draggable dialogs
  - Dialog components are now outside of a drag and drop context
- Fixed layout issues regarding the buttons on the board header
- Fixed layout issues regarding board cards (#43)
- Fixed an issue where columns where expanding its height while dragging (#45)
- Fixed vote count dialog (#42)
  - Disallowed negative vote counts
  - Cancel button should work properly now

## [0.3.0] - 2019-08-22

### Added

- **Validation** for all Dialogs
- **Error messages** for all text fields that are invalid
- **ESLint** for `backend`
- **Roles**
  - Users are now divided into roles: **1 moderator and N participants**
  - Moderator

playwright.config.ts
import { defineConfig, devices } from "@playwright/test";

/**
 * See https://playwright.dev/docs/test-configuration.
 */
export default defineConfig({
  testDir: "./e2e",
  /* Maximum time one test can run for. */
  timeout: 30 * 1000,
  expect: {
    /**
     * Maximum time expect() should wait for the condition to be met.
     * For example in `await expect(locator).toHaveText();`
     */
    timeout: 5000,
  },
  /* Run tests in files in parallel */
  fullyParallel: true,
  /* Fail the build on CI if you accidentally left test.only in the source code. */
  forbidOnly: !!process.env.CI,
  /* Retry on CI only */
  retries: process.env.CI ? 1 : 0,
  /* Opt out of parallel tests on CI. */
  workers: process.env.CI ? 1 : undefined,
  reporter: process.env.CI ? "github" : "list",
  webServer: [
    {
      command: "npm run frontend:dev",
      port: 3000,
      timeout: 120 * 1000,
      reuseExistingServer: !process.env.CI,
    },
    {
      command: "npm run backend:dev",
      port: 3001,
      timeout: 120 * 1000,
      reuseExistingServer: !process.env.CI,
    },
  ],
  use: {
    /* Maximum time each action such as `click()` can take. Defaults to 0 (no limit). */
    actionTimeout: 0,
    trace: "on-first-retry",
    /* Base URL to use in actions like `await page.goto('/')`. */
    baseURL: "http://localhost:3000/",
  },

  /* Configure projects for major browsers */
  projects: [
    {
      name: "chromium",
      use: { ...devices["Desktop Chrome"], permissions: ["clipboard-read"] },
    },
    {
      name: "firefox",
      use: { ...devices["Desktop Firefox"] },
    },
  ],

  /* Folder for test artifacts such as screenshots, videos, traces, etc. */
  // outputDir: 'test-results/',

  /* Run your local dev server before starting the tests */
  // webServer: {
  //   command: 'npm run start',
  //   port: 3000,
  // },
});
